protocols/autonat/src/protocol.rs

```diff
@@ -0,0 +1,304 @@
+                maddrs.push(maddr);
+            }
+            maddrs
+        };
+        Ok(Self { peer_id, addrs })
+    }
+
+    pub fn into_bytes(self) -> Vec<u8> {
+        let peer_id = self.peer_id.to_bytes();
+        let addrs = self.addrs.into_iter().map(|addr| addr.to_vec()).collect();
+
+        let msg = structs_proto::Message {
+            r#type: Some(structs_proto::message::MessageType::Dial as _),
+            dial: Some(structs_proto::message::Dial {
+                peer: Some(structs_proto::message::PeerInfo {
+                    id: Some(peer_id),
+                    addrs,
+                }),
+            }),
+            dial_response: None,
+        };
+
+        let mut bytes = Vec::with_capacity(msg.encoded_len());
+        msg.encode(&mut bytes)
+            .expect("Vec<u8> provides capacity as needed");
+        bytes
+    }
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum ResponseError {
+    DialError,
+    DialRefused,
+    BadRequest,
+    InternalError,
+}
+
+impl From<ResponseError> for i32 {
+    fn from(t: ResponseError) -> Self {
+        match t {
+            ResponseError::DialError => 100,
+            ResponseError::DialRefused => 101,
+            ResponseError::BadRequest => 200,
+            ResponseError::InternalError => 300,
+        }
+    }
+}
+
+impl TryFrom<i32> for ResponseError {
+    type Error = io::Error;
+
+    fn try_from(value: i32) -> Result<Self, Self::Error> {
+        match value {
+            100 => Ok(ResponseError::DialError),
+            101 => Ok(ResponseError::DialRefused),
+            200 => Ok(ResponseError::BadRequest),
+            300 => Ok(ResponseError::InternalError),
+            _ => Err(io::Error::new(
+                io::ErrorKind::InvalidData,
+                "invalid response error type",
+            )),
+        }
+    }
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum DialResponse {
+    Ok(Multiaddr),
+    Err(ResponseError),
+}
+
+impl DialResponse {
+    pub fn from_bytes(bytes: &[u8]) -> Result<Self, io::Error> {
+        let msg = structs_proto::Message::decode(bytes)
+            .map_err(|err| io::Error::new(io::ErrorKind::InvalidData, err))?;
+        if msg.r#type != Some(structs_proto::message::MessageType::DialResponse as _) {
+            return Err(io::Error::new(io::ErrorKind::InvalidData, "invalid type"));
+        }
+
+        Ok(match msg.dial_response {
+            Some(structs_proto::message::DialResponse {
+                status: Some(0),
+                status_text: None,
+                addr: Some(addr),
+            }) => {
+                let addr = Multiaddr::try_from(addr)
+                    .map_err(|err| io::Error::new(io::ErrorKind::InvalidData, err))?;
+                Self::Ok(addr)
+            }
+            Some(structs_proto::message::DialResponse {
+                status: Some(status),
+                status_text: _,
```

Comment:
I think `status_text` is super helpful for debugging. How about either including it in `ResponseError` or at least logging it on `log::debug`?

Reply:
Added with 2f23e4b871421dd4f40241e7b4df2f27862c7c77 and b70428e28c4d2a2315598deeb7ebd43838473ac8.
Since throughout the rest of the library afaik it's not common to return strings as error message, I just logged it but not included it it in the `NatStatus`. But I am rather neutral on this point, so I could also add it to `ResponseError` if you prefer that.